File: src/shared/uiKit/chart/universalChart/plot/line/autoLine/monotoneXPath.zig
const std = @import("std");

// [BUILD]
// zig build-exe monotoneXPath.zig -target wasm32-freestanding -fno-entry -rdynamic -O ReleaseSmall

const allocator = std.heap.wasm_allocator;

var empty_bytes: [0]u8 = .{};
var result_buffer: []u8 = empty_bytes[0..];
var result_length: usize = 0;

const NUMBER_CAPACITY = 24;
const MOVE_CAPACITY = 2 + NUMBER_CAPACITY + 1 + NUMBER_CAPACITY;
const SEGMENT_CAPACITY = 2 + NUMBER_CAPACITY + 1 + NUMBER_CAPACITY + 2 + NUMBER_CAPACITY + 1 + NUMBER_CAPACITY + 2 + NUMBER_CAPACITY + 1 + NUMBER_CAPACITY + 1;

const PathWriter = struct {
    buffer: []u8,
    pos: usize,

    fn init(capacity: usize) !PathWriter {
        if (result_buffer.len == 0) {
            result_buffer = try allocator.alloc(u8, capacity);
        } else if (result_buffer.len < capacity) {
            result_buffer = try allocator.realloc(result_buffer, capacity);
        }

        return .{
            .buffer = result_buffer,
            .pos = 0,
        };
    }

    fn deinit(self: *PathWriter) void {
        if (self.buffer.len > 0) {
            allocator.free(self.buffer);
        }

        result_buffer = empty_bytes[0..];
        result_length = 0;

        self.buffer = empty_bytes[0..];
        self.pos = 0;
    }

    fn finish(self: *PathWriter) void {
        result_buffer = self.buffer;
        result_length = self.pos;
    }

    fn ensure(self: *PathWriter, extra: usize) !void {
        if (self.pos + extra <= self.buffer.len) return;

        var next_len = @max(self.buffer.len * 2, self.pos + extra);
        if (next_len == 0) next_len = extra;

        self.buffer = try allocator.realloc(self.buffer, next_len);
        result_buffer = self.buffer;
    }

    fn byte(self: *PathWriter, value: u8) void {
        self.buffer[self.pos] = value;
        self.pos += 1;
    }

    fn text(self: *PathWriter, value: []const u8) void {
        @memcpy(self.buffer[self.pos .. self.pos + value.len], value);
        self.pos += value.len;
    }

    fn uint(self: *PathWriter, value: u64) void {
        if (value == 0) {
            self.byte('0');
            return;
        }

        var digits: [20]u8 = undefined;
        var count: usize = 0;
        var current = value;

        while (current > 0) {
            digits[count] = '0' + @as(u8, @intCast(current % 10));
            current /= 10;
            count += 1;
        }

        while (count > 0) {
            count -= 1;
            self.byte(digits[count]);
        }
    }

    fn number2(self: *PathWriter, value: f64) !void {
        if (!std.math.isFinite(value)) return error.InvalidNumber;

        const scaled_float = @floor(value * 100.0 + 0.5);
        const max_i64: f64 = @floatFromInt(std.math.maxInt(i64));
        const min_i64: f64 = @floatFromInt(std.math.minInt(i64));

        if (scaled_float > max_i64 or scaled_float < min_i64) {
            return error.InvalidNumber;
        }

        const scaled: i64 = @intFromFloat(scaled_float);

        if (scaled == 0) {
            self.byte('0');
            return;
        }

        var abs_scaled: u64 = undefined;

        if (scaled < 0) {
            self.byte('-');
            abs_scaled = @as(u64, @intCast(-(scaled + 1))) + 1;
        } else {
            abs_scaled = @intCast(scaled);
        }

        const int_part = abs_scaled / 100;
        const frac: u8 = @intCast(abs_scaled % 100);

        self.uint(int_part);

        if (frac == 0) return;

        self.byte('.');

        const tens = frac / 10;
        const ones = frac - tens * 10;

        self.byte('0' + tens);

        if (ones != 0) {
            self.byte('0' + ones);
        }
    }
};

export fn allocF64(len: usize) usize {
    const slice = allocator.alloc(f64, len) catch return 0;
    return @intFromPtr(slice.ptr);
}

export fn freeF64(ptr: usize, len: usize) void {
    if (ptr == 0 or len == 0) return;

    const slice = @as([*]f64, @ptrFromInt(ptr))[0..len];
    allocator.free(slice);
}

export fn resultPtr() usize {
    if (result_length == 0) return 0;
    return @intFromPtr(result_buffer.ptr);
}

export fn resultLen() usize {
    return result_length;
}

export fn buildMonotoneXPath(
    x_ptr: usize,
    y_ptr: usize,
    len: usize,
    smoothing: f64,
    bounds_min_x: f64,
    bounds_max_x: f64,
    bounds_min_y: f64,
    bounds_max_y: f64,
    layout_min_x: f64,
    layout_max_x: f64,
    layout_min_y: f64,
    layout_max_y: f64,
    visible_min_x: f64,
    visible_max_x: f64,
    visible_min_y: f64,
    visible_max_y: f64,
) u32 {
    clearResult();

    if (x_ptr == 0 or y_ptr == 0 or len == 0) return 0;
    if (!allFinite(.{
        smoothing,
        bounds_min_x,
        bounds_max_x,
        bounds_min_y,
        bounds_max_y,
        layout_min_x,
        layout_max_x,
        layout_min_y,
        layout_max_y,
        visible_min_x,
        visible_max_x,
        visible_min_y,
        visible_max_y,
    })) return 0;

    const input_x = @as([*]const f64, @ptrFromInt(x_ptr))[0..len];
    const input_y = @as([*]const f64, @ptrFromInt(y_ptr))[0..len];

    const w = bounds_max_x - bounds_min_x;
    const h = bounds_max_y - bounds_min_y;
    const layout_w = layout_max_x - layout_min_x;
    const layout_h = layout_max_y - layout_min_y;

    var visible_start = len;
    var visible_end: usize = 0;

    if (len == 1) {
        if (input_x[0] >= visible_min_x and input_x[0] <= visible_max_x and input_y[0] >= visible_min_y and input_y[0] <= visible_max_y) {
            visible_start = 0;
            visible_end = 0;
        }
    } else {
        for (0..(len - 1)) |i| {
            const seg_min_x = @min(input_x[i], input_x[i + 1]);
            const seg_max_x = @max(input_x[i], input_x[i + 1]);
            if (seg_max_x < visible_min_x or seg_min_x > visible_max_x) continue;

            const seg_min_y = @min(input_y[i], input_y[i + 1]);
            const seg_max_y = @max(input_y[i], input_y[i + 1]);
            if (seg_max_y < visible_min_y or seg_min_y > visible_max_y) continue;

            if (visible_start == len) visible_start = i;
            visible_end = i + 1;
        }
    }

    if (visible_start == len) return 1;

    if (visible_start > 0) visible_start -= 1;
    if (visible_end + 1 < len) visible_end += 1;

    const visible_len = visible_end - visible_start + 1;

    const x = allocator.alloc(f64, visible_len) catch return 0;
    defer allocator.free(x);

    const y = allocator.alloc(f64, visible_len) catch return 0;
    defer allocator.free(y);

    for (0..visible_len) |i| {
        const source_index = visible_start + i;
        x[i] = (input_x[source_index] - bounds_min_x) / w * layout_w + layout_min_x;
        y[i] = (bounds_max_y - input_y[source_index]) / h * layout_h + layout_min_y;

        if (!std.math.isFinite(x[i]) or !std.math.isFinite(y[i])) return 0;
    }

    if (visible_len == 1) {
        var writer_single = PathWriter.init(@max(@as(usize, 64), MOVE_CAPACITY)) catch return 0;
        emitMovePath(&writer_single, x[0], y[0]) catch {
            writer_single.deinit();
            return 0;
        };
        writer_single.finish();
        return 1;
    }

    if (visible_len == 2) {
        var writer_line = PathWriter.init(@max(@as(usize, 96), MOVE_CAPACITY + SEGMENT_CAPACITY)) catch return 0;
        emitLinePath(&writer_line, x[0], y[0], x[1], y[1]) catch {
            writer_line.deinit();
            return 0;
        };
        writer_line.finish();
        return 1;
    }

    const segment_count = visible_len - 1;

    const dx = allocator.alloc(f64, segment_count) catch return 0;
    defer allocator.free(dx);

    const slope = allocator.alloc(f64, segment_count) catch return 0;
    defer allocator.free(slope);

    const tangent = allocator.alloc(f64, visible_len) catch return 0;
    defer allocator.free(tangent);

    for (0..segment_count) |i| {
        dx[i] = x[i + 1] - x[i];
        const dy = y[i + 1] - y[i];
        const denom = if (dx[i] == 0.0) 1.0 else dx[i];
        slope[i] = dy / denom;

        if (!std.math.isFinite(slope[i])) return 0;
    }

    tangent[0] = slope[0];
    tangent[visible_len - 1] = slope[segment_count - 1];

    for (1..segment_count) |i| {
        const s0 = slope[i - 1];
        const s1 = slope[i];

        tangent[i] = if (s0 * s1 <= 0.0) 0.0 else (s0 + s1) / 2.0;
    }

    for (0..segment_count) |i| {
        if (slope[i] == 0.0) {
            tangent[i] = 0.0;
            tangent[i + 1] = 0.0;
        } else {
            const a = tangent[i] / slope[i];
            const b = tangent[i + 1] / slope[i];
            const hypotenuse = @sqrt(a * a + b * b);

            if (hypotenuse > 3.0) {
                const t = 3.0 / hypotenuse;
                tangent[i] = t * a * slope[i];
                tangent[i + 1] = t * b * slope[i];
            }
        }
    }

    const k = if (smoothing < 0.0) 0.0 else if (smoothing > 1.0) 1.0 else smoothing;

    var writer = PathWriter.init(@max(@as(usize, 128), visible_len * 96)) catch return 0;

    emitPath(&writer, x, y, dx, tangent, k) catch {
        writer.deinit();
        return 0;
    };

    writer.finish();

    return 1;
}

fn clearResult() void {
    result_length = 0;
}

fn allFinite(values: anytype) bool {
    inline for (values) |value| {
        if (!std.math.isFinite(value)) return false;
    }

    return true;
}

fn emitPath(writer: *PathWriter, x: []const f64, y: []const f64, dx: []const f64, tangent: []const f64, k: f64) !void {
    try writer.ensure(MOVE_CAPACITY);
    writer.text("M ");
    try writer.number2(x[0]);
    writer.byte(' ');
    try writer.number2(y[0]);

    for (0..dx.len) |i| {
        try writer.ensure(SEGMENT_CAPACITY);

        const cp1x = x[i] + (dx[i] / 3.0) * k;
        const cp1y = y[i] + ((tangent[i] * dx[i]) / 3.0) * k;

        const cp2x = x[i + 1] - (dx[i] / 3.0) * k;
        const cp2y = y[i + 1] - ((tangent[i + 1] * dx[i]) / 3.0) * k;

        writer.text("C ");
        try writer.number2(cp1x);
        writer.byte(' ');
        try writer.number2(cp1y);
        writer.text(", ");
        try writer.number2(cp2x);
        writer.byte(' ');
        try writer.number2(cp2y);
        writer.text(", ");
        try writer.number2(x[i + 1]);
        writer.byte(' ');
        try writer.number2(y[i + 1]);
        writer.byte(' ');
    }
}

fn emitMovePath(writer: *PathWriter, x: f64, y: f64) !void {
    try writer.ensure(MOVE_CAPACITY);
    writer.text("M ");
    try writer.number2(x);
    writer.byte(' ');
    try writer.number2(y);
}

fn emitLinePath(writer: *PathWriter, x0: f64, y0: f64, x1: f64, y1: f64) !void {
    try writer.ensure(MOVE_CAPACITY + SEGMENT_CAPACITY);
    writer.text("M ");
    try writer.number2(x0);
    writer.byte(' ');
    try writer.number2(y0);
    writer.text(" L ");
    try writer.number2(x1);
    writer.byte(' ');
    try writer.number2(y1);
}

File: src/shared/uiKit/chart/universalChart/plot/line/autoLine/MonotoneXPath.ts
import wasmInit from './monotoneXPath/monotoneXPath.wasm?init'

type MonotoneXPathWasm = {
  memory: WebAssembly.Memory
  allocF64: (len: number) => number
  freeF64: (ptr: number, len: number) => void
  buildMonotoneXPath: (
    xPtr: number,
    yPtr: number,
    len: number,
    smoothing: number,
    boundsMinX: number, boundsMaxX: number, boundsMinY: number, boundsMaxY: number,
    layoutMinX: number, layoutMaxX: number, layoutMinY: number, layoutMaxY: number,
    visibleMinX: number, visibleMaxX: number, visibleMinY: number, visibleMaxY: number
  ) => number
  resultPtr: () => number
  resultLen: () => number
}

let monotoneXPathWasm: MonotoneXPathWasm | null = null

wasmInit()
  .then((instance) => monotoneXPathWasm = instance.exports as unknown as MonotoneXPathWasm)
  .catch(() => monotoneXPathWasm = null)

const pathDecoder = new TextDecoder()

type Bounds = {
  minX: number,
  maxX: number,
  minY: number,
  maxY: number
}

export class MonotoneXPath {

  private x: Float64Array
  private y: Float64Array
  private wasm: MonotoneXPathWasm | null = null
  private wasmXPtr = 0
  private wasmYPtr = 0

  constructor(points: { x: number, y: number }[]) {
    const n = points.length
    this.x = new Float64Array(n)
    this.y = new Float64Array(n)

    for (let i = 0; i < n; i++) {
      this.x[i] = points[i].x
      this.y[i] = points[i].y
    }
  }

  getPath(smoothing = 1, bounds: Bounds, layout: Bounds, visibleLayout: Bounds): string {
    const n = this.x.length

    if (n === 0) return ''

    const visibleBounds = getVisibleChartBounds(bounds, layout, visibleLayout)

    const wasm = monotoneXPathWasm

    if (wasm) {
      try {
        if (this.ensureWasmBuffers(wasm)) {

          const ok = wasm.buildMonotoneXPath(
            this.wasmXPtr, this.wasmYPtr, n, smoothing,
            bounds.minX, bounds.maxX, bounds.minY, bounds.maxY,
            layout.minX, layout.maxX, layout.minY, layout.maxY,
            visibleBounds.minX, visibleBounds.maxX, visibleBounds.minY, visibleBounds.maxY
          )

          if (ok) {
            const ptr = wasm.resultPtr()
            const len = wasm.resultLen()
            const res = pathDecoder.decode(new Uint8Array(wasm.memory.buffer, ptr, len))
            return res
          }
        }
      } catch {
        this.freeWasmBuffers()
      }
    }

    return monotoneXPathFallback({
      x: this.x,
      y: this.y,
      smoothing: smoothing,
      bounds: bounds,
      layout: layout,
      visibleBounds: visibleBounds
    })
  }

  dispose() {
    this.freeWasmBuffers()
    this.x = new Float64Array(0)
    this.y = new Float64Array(0)
  }

  private ensureWasmBuffers(wasm: MonotoneXPathWasm): boolean {
    if (this.wasm === wasm && this.wasmXPtr !== 0 && this.wasmYPtr !== 0) return true

    this.freeWasmBuffers()

    const n = this.x.length
    const xPtr = wasm.allocF64(n)
    const yPtr = wasm.allocF64(n)

    if (xPtr === 0 || yPtr === 0) {
      if (xPtr !== 0) wasm.freeF64(xPtr, n)
      if (yPtr !== 0) wasm.freeF64(yPtr, n)
      return false
    }

    new Float64Array(wasm.memory.buffer, xPtr, n).set(this.x)
    new Float64Array(wasm.memory.buffer, yPtr, n).set(this.y)

    this.wasm = wasm
    this.wasmXPtr = xPtr
    this.wasmYPtr = yPtr

    return true
  }

  private freeWasmBuffers() {
    const wasm = this.wasm

    if (wasm) {
      try {
        if (this.wasmXPtr !== 0) wasm.freeF64(this.wasmXPtr, this.x.length)
        if (this.wasmYPtr !== 0) wasm.freeF64(this.wasmYPtr, this.y.length)
      } catch {
        // Ignore allocator failures and let the instance fall back to JS.
      }
    }

    this.wasm = null
    this.wasmXPtr = 0
    this.wasmYPtr = 0
  }

}

function fmt2(value: number) {
  return Math.round(value * 100) / 100
}

function monotoneXPathFallback(options: {
  x: Float64Array,
  y: Float64Array,
  smoothing: number,
  bounds: Bounds,
  layout: Bounds,
  visibleBounds: Bounds
}): string {

  const { smoothing, bounds, layout, visibleBounds } = options

  const w = bounds.maxX - bounds.minX
  const h = bounds.maxY - bounds.minY

  const lW = layout.maxX - layout.minX
  const lH = layout.maxY - layout.minY

  const transformX = (x: number) => (x - bounds.minX) / w * lW + layout.minX
  const transformY = (y: number) => (bounds.maxY - y) / h * lH + layout.minY

  const visibleRange = getVisiblePointRange(options.x, options.y, visibleBounds)
  if (!visibleRange) return ''

  const n = visibleRange.end - visibleRange.start + 1
  const x = new Array<number>(n)
  const y = new Array<number>(n)

  for (let i = 0; i < n; i++) {
    const sourceIndex = visibleRange.start + i
    x[i] = transformX(options.x[sourceIndex])
    y[i] = transformY(options.y[sourceIndex])
  }

  if (n === 1) return `M${fmt2(x[0])} ${fmt2(y[0])}`
  if (n === 2) return `M${fmt2(x[0])} ${fmt2(y[0])} L${fmt2(x[1])} ${fmt2(y[1])}`


  const dx = new Array(Math.max(0, n - 1))
  const slope = new Array(Math.max(0, n - 1))
  const tangent = new Array(n)

  for (let i = 0; i < n - 1; i++) {
    dx[i] = x[i + 1] - x[i]
    const dy = y[i + 1] - y[i]
    slope[i] = dy / (dx[i] || 1)
  }

  tangent[0] = slope[0]
  tangent[n - 1] = slope[n - 2]

  for (let i = 1; i < n - 1; i++) {
    const s0 = slope[i - 1]
    const s1 = slope[i]

    if (s0 * s1 <= 0) {
      tangent[i] = 0
    } else {
      tangent[i] = (s0 + s1) / 2
    }
  }

  // Prevent overshoot.
  for (let i = 0; i < n - 1; i++) {
    if (slope[i] === 0) {
      tangent[i] = 0
      tangent[i + 1] = 0
    } else {
      const a = tangent[i] / slope[i]
      const b = tangent[i + 1] / slope[i]
      const h = Math.hypot(a, b)

      if (h > 3) {
        const t = 3 / h
        tangent[i] = t * a * slope[i]
        tangent[i + 1] = t * b * slope[i]
      }
    }
  }

  const k = Math.max(0, Math.min(1, smoothing))

  let d = ''
  d += `M ${fmt2(x[0])} ${fmt2(y[0])}`

  for (let i = 0; i < n - 1; i++) {
    const p0 = { x: x[i], y: y[i] }
    const p1 = { x: x[i + 1], y: y[i + 1] }

    const cp1x = p0.x + (dx[i] / 3) * k
    const cp1y = p0.y + ((tangent[i] * dx[i]) / 3) * k

    const cp2x = p1.x - (dx[i] / 3) * k
    const cp2y = p1.y - ((tangent[i + 1] * dx[i]) / 3) * k

    d += `C ${fmt2(cp1x)} ${fmt2(cp1y)}, ${fmt2(cp2x)} ${fmt2(cp2y)}, ${fmt2(p1.x)} ${fmt2(p1.y)} `
  }

  return d
}

function getVisibleChartBounds(bounds: Bounds, layout: Bounds, visibleLayout: Bounds): Bounds {
  const width = bounds.maxX - bounds.minX
  const height = bounds.maxY - bounds.minY
  const layoutWidth = layout.maxX - layout.minX
  const layoutHeight = layout.maxY - layout.minY

  const layoutToChartX = (x: number) => bounds.minX + (x - layout.minX) / layoutWidth * width
  const layoutToChartY = (y: number) => bounds.maxY - (y - layout.minY) / layoutHeight * height

  return {
    minX: layoutToChartX(visibleLayout.minX),
    maxX: layoutToChartX(visibleLayout.maxX),
    minY: layoutToChartY(visibleLayout.maxY),
    maxY: layoutToChartY(visibleLayout.minY)
  }
}

function getVisiblePointRange(x: Float64Array, y: Float64Array, visibleBounds: Bounds) {
  const n = x.length

  if (n === 1) {
    if (x[0] < visibleBounds.minX || x[0] > visibleBounds.maxX || y[0] < visibleBounds.minY || y[0] > visibleBounds.maxY) return null
    return { start: 0, end: 0 }
  }

  let start = -1
  let end = -1

  for (let i = 0; i < n - 1; i++) {
    const segMinX = Math.min(x[i], x[i + 1])
    const segMaxX = Math.max(x[i], x[i + 1])
    if (segMaxX < visibleBounds.minX || segMinX > visibleBounds.maxX) continue

    const segMinY = Math.min(y[i], y[i + 1])
    const segMaxY = Math.max(y[i], y[i + 1])
    if (segMaxY < visibleBounds.minY || segMinY > visibleBounds.maxY) continue

    if (start === -1) start = i
    end = i + 1
  }

  if (start === -1 || end === -1) return null

  return {
    start: Math.max(0, start - 1),
    end: Math.min(n - 1, end + 1)
  }
}